chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+             morphed_conditions)
+         )
+      )
+
+    (defun-inline coin_info_for_coin_bundle (coin_bundle)
+      (f coin_bundle)
+    )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;; lineage checking
+
+     ;; return true iff parent of `this_coin_info` is provably a cat
+     (defun is_parent_cat (
+       cat_mod_struct
+       parent_id
+       (parent_parent_coin_id parent_inner_puzzle_hash parent_amount)
+     )
+
+       (if (= parent_id
+              (sha256 parent_parent_coin_id
+                      (cat_puzzle_hash cat_mod_struct parent_inner_puzzle_hash)
+                      parent_amount
+              )
+           )
+           1
+           (x)
+       )
+     )
+
+     ;; return true iff the lineage proof is valid
+     ;; lineage_proof is of one of two forms:
+     ;;  (1 . (parent_parent_coin_id parent_inner_puzzle_hash parent_amount))
+     ;;  (0 . some_opaque_proof_passed_to_GENESIS_COIN_CHECKER)
+     ;; so the `f` value determines what kind of proof it is, and the `r` value is the proof
+
+     (defun-inline limitations_program_from_cat_struct (cat_mod_struct)
+       (f (r (r cat_mod_struct)))
+     )
+
+     (defun-inline limitations_program_hash_from_cat_struct (cat_mod_struct)
+       (f (r (r (r cat_mod_struct))))
+     )
+
+    (defun is_bundle_valid ((coin_info . lineage_proof) limitations_program Truths extra_delta inner_conditions)
+      (if limitations_program
+          (assert (= (sha256tree1 limitations_program) (limitations_program_hash_from_cat_struct (cat_struct_truth Truths)))
+            (a limitations_program
+                (list
+                  Truths
+                  extra_delta
+                  (if (f lineage_proof)
+                      (is_parent_cat (cat_struct_truth Truths) (my_parent_cat_truth Truths) (r lineage_proof))
+                      ()
+                  )
+                  inner_conditions
+                  (r lineage_proof)
+                )
+            )
+          )
+          (assert (is_parent_cat (cat_struct_truth Truths) (my_parent_cat_truth Truths) lineage_proof)
+            inner_conditions
+          )
+      )
+    )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+     (defun stager_three (
+         Truths
+         inner_conditions
+         prev_coin_id
+         this_coin_bundle
+         next_coin_id
+         prev_subtotal
+         extra_delta
+      )
+      (is_bundle_valid
+            this_coin_bundle
+            (limitations_program_from_cat_struct (cat_struct_truth Truths))
+            Truths
+            extra_delta
+            (generate_final_output_conditions
+              prev_subtotal
+              ; the expression on the next line calculates `this_subtotal` by adding the delta to `prev_subtotal`
+              (+ (+ prev_subtotal (- (input_amount_for_coin (coin_info_for_coin_bundle this_coin_bundle)) (output_totals inner_conditions))) extra_delta)
```

Comment:
You don't need a second `+` operator here, since `(+ (+ A B) C)` = `(+ A B C)`. It'll be easier to read too.
```suggestion
              (+ prev_subtotal (- (input_amount_for_coin (coin_info_for_coin_bundle this_coin_bundle)) (output_totals inner_conditions)) extra_delta)
```

Reply:
Done